### ISSUE TITLE
Roadmap to DiffSharp 1.0

### ISSUE THREAD

**Issue by dsyme** [DiffSharp/DiffSharp#69]
This is a catalog issue to track what needs doing for DiffSharp 1.0, based on 1:1 discussions with @gbaydin.  There will be a long list of other things, we'll extend this as necessary

* [x] Typed Backend.None CPU tensors ([draft](https://github.com/DiffSharp/DiffSharp/pull/77))

* [x] Add `keep_dims` on `Mean` ([done](https://github.com/DiffSharp/DiffSharp/pull/70))

* [x] Fix `CompareTo` in RawTensorFloat32CPU.fs ([done](https://github.com/DiffSharp/DiffSharp/pull/84))

* [x] Broadcasting. Full pytorch-style broadcasting for `Add` (see [TODO here](https://github.com/DiffSharp/DiffSharp/blob/9aa23cff72b8b8eb2cb0ebbf2973a8d892ad186b/src/DiffSharp.Core/Tensor.fs#L278)).   The design principle expected here is "we should do the same thing as PyTorch".  Similarly full pytorch-style broadcasting for `Mul` and other operations. ([done](https://github.com/DiffSharp/DiffSharp/pull/85))
 
* [x] Convolutions (@gbaydin)

* [x] Transposed convolutions

* [x] Batchnorm

* [x] Dropout

* [x] Switch to Python-style casing on all operations to align with SciSharp

* [x] Remove excess overloads and use optional arguments instead

* [x] Finalize API `dsharp.abc`

* [x] libtorch and cuda backends

* [x] Add `Reshape` (similar code to `View`)

* [x] Add `OneHot`

* [x] Differentiation API

* [x] Optimizers

* [x] Fix possible memory leak on Linux

* [x] Tensor save/load

* [x] General transpose 

* [x] probability distributions

* [x] Generalization and batching of `Transpose`.

* [x] Zero-size tensors #150

* [x] Docs tooling #134

* [x] Docs #167

* [x] PyTorch Half support

* [x] Batching conv1d/2d/3d/... https://github.com/DiffSharp/DiffSharp/issues/98

* [x] Batching for `MatMul`.  Currently [no batching is supported, only 2D x 2D](https://github.com/DiffSharp/DiffSharp/blob/9aa23cff72b8b8eb2cb0ebbf2973a8d892ad186b/src/DiffSharp.Core/Tensor.fs#L434) [done](https://github.com/DiffSharp/DiffSharp/pull/88)

* [x] Einstein summation #92 

* [ ] norm #93 

* [ ] matrix inverse

Things out of scope for 1.0

* [ ] Strided views

* [ ] Quantized 

* [ ] Complex

* [ ] Sparse

**Comment by kevmal** [comment]
> Proposal to switch to Python-style casing on all operations to align with SciSharp
Not sure what the motivation is on this one?

**Comment by smoothdeveloper** [comment]
@kevmal when I look at python code, squinting a bit, it is almost like procedural F# code.

I guess, the appeal will be to show F#/DiffSharp code to pythoners and tell "see all the same, but a compiler and tooling which has your back".

Once enough pythoners see this, and enough libraries cover the basics of SciSharp and others in the ML stack, we'll see skyrocketing F# adoption.

this looks like the "???" step in:

* step 1 : tell pythoners about F# 
* step 2 : ???
* step 3 : make pythoners squint 
* step 4 : profit !!!!

https://knowyourmeme.com/memes/profit

**Comment by dsyme** [comment]
Note that the direction of travel here is that the DiffSharp 1.0 API looks and acts a lot like an F# version of the PyTorch API.

**Comment by dsyme** [comment]
> Not sure what the motivation is on this one?

It's open for discussion and would need to be trialled but I've been pretty convinced that SciSharp is doing the right thing here.  

99.9% of people learning deep learning do it through Python, and thatt will continue for the next 5 year. To get people to even consider using F#/.NET then learning barrier must be as low as possible.    This includes documentation. In Tensorflow.NET the SciSharp people have made a .NET entry to TF that is very easy and natural to get going with, and naming is a big part of that.

**Comment by dsyme** [comment]
Gunes has thought a lot about these things too and we need to discuss - we only chatted about it briefly today.

**Comment by kevmal** [comment]
> > Not sure what the motivation is on this one?
> 
> It's open for discussion and would need to be trialled but I've been pretty convinced that SciSharp is doing the right thing here.
> 
> 99.9% of people learning deep learning do it through Python, and thatt will continue for the next 5 year. To get people to even consider using F#/.NET then learning barrier must be as low as possible. This includes documentation. In Tensorflow.NET the SciSharp people have made a .NET entry to TF that is very easy and natural to get going with, and naming is a big part of that.

Noticed it's at least not snake_case (i.e. zerosLike not zeros_like) since it I found it really doesn't seem to work well with tooling.

But speaking as someone with a math background who came from MATLAB/python/R prior to F#... I'd say case won't make much of a difference except seem inconsistent. Furthermore I know what `randn` is because of MATLAB, what `RandomNormal` is is just obvious regardless of if I have MATLAB experience or not. So why make it more difficult for those not coming from python? Not to mention one of the advantages to F# is the tooling and with VS2017 (VS2019 is unfortunately too slow) I'd have various libraries referenced and without touching a web browser if I need an ADF test:

![image](https://user-images.githubusercontent.com/7281367/75904535-9f122980-5e00-11ea-8009-1979390cc3f1.png)

Now if I need to sample from a Gaussian distribution with no prior knowledge? There's a chance I'd figure it out after typing "rand" but also a decent chance I'd have to go search to learn about `randn`.

With that said, `randn` is just an example and my main point is that "because that's how it is in python or library X" is an unfortunate reason. Focus on the best F# experience.

The choice SciSharp made is great in the context of the amount of python code that needs to be ported to .NET for a "complete" TF experience. Code can just be copied, some updates and it's good to go. Given the target demo of those who actually want to build and train models I think the advantage somewhat stops there. Since I would assume that demo has knowledge with the core concepts and would like to build out their own ideas in the cleanest easiest way possible.


 


> @kevmal when I look at python code, squinting a bit, it is almost like procedural F# code.
> 
> I guess, the appeal will be to show F#/DiffSharp code to pythoners and tell "see all the same, but a compiler and tooling which has your back".
> 
> Once enough pythoners see this, and enough libraries cover the basics of SciSharp and others in the ML stack, we'll see skyrocketing F# adoption.
> 
> this looks like the "???" step in:
> 
> * step 1 : tell pythoners about F#
> * step 2 : ???
> * step 3 : make pythoners squint
> * step 4 : profit !!!!
> 
> https://knowyourmeme.com/memes/profit

I feel like if it were that simple everyone would be using Julia. Seems like there's only advantages to Julia over python (okay there is some first-call lag due to compilation), not to mention seamless python interop if needed. But with the conversations I've had it seems most are quite content with the python experience. I think if it needs to look like python before being considered you already lost the sale. 

Again I think focus on the best F# experience, if the F# community does interesting/unique things with the result then interest may spike. Though the same argument holds here with julia, look at the interesting things being done with zygote.jl and flux.jl. I still think it's the best strategy.

**Comment by dsyme** [comment]
@kevmal I know what you're saying, but I think the general detailed design alignment with PyTorch is right at this point.  Given that, name alignment following SciSharp's model seems reasonable too.  

It could be it's an issue revisted lately, or with a secondary API, v2.0 or fork aligning with .NET naming conventions.  But at the moment I'm personally finding it highly clarifying to be aligning the subtle design choices and decisions with PyTorch.

Another factor is that we're vaguely interested in bringing DiffSharp's capabilities into the PyTorch ecosystem, potentially via a Python package (reimplemented in Python) so direct comparison is useful for us there.

**Comment by pkese** [comment]
I'm not in favor of the new naming. While I agree with following PyTorch naming convetions, we should also respect our own F# casing convetions.

So couldn't we just explain Pythonistas this one sentence

_Everything is the same as in PyTorch, except in F# we write object methods in PascalCase like `vec.Length` and static functions in snakeCase like `Array.length vec`_

...and avoid the damage to the F# side.  

I think, sooner or later even the potential Pythonista converts will have to interact with the rest of the F# ecosystem and find such casing inconsistencies weird. Let's not damage the F# API just for the 'profit' meme.

**Comment by dsyme** [comment]
I think it's fair to say existing F# programmers are not the majority of the (currently hypothetical) target users of DiffSharp.

**Comment by dsyme** [comment]
Remaining work listed here is now covered by the "1.0" tag
